widget/class-widget.php

```diff
@@ -0,0 +1,229 @@
+      parent::__construct(
+         'WI_Volunteer_Management_Widget', // Base ID
+         __('Volunteer Management Widget', 'text_domain'), // Name
+         array( 'description' => __( 'A list of volunteer opportunities.', 'text_domain' ), ) // Widget Options
+      );
+
+   }
+
+   /**
+   * Front-end display of widget.
+   *
+   * @see WP_Widget::widget()
+   *
+   * @param array $args     Widget arguments.
+   * @param array $instance Saved values from database.
+   */
+   public function widget( $args, $instance ) {
+
+      // Global variables to  be used in templates/opps-list-widget.php
+      global $display_opp_when;
+      global $display_opp_where;
+      global $display_opp_spots;
+
+      isset( $instance['opp_info_when'] ) ? $display_opp_when = true : $display_opp_when = false;
+      isset( $instance['opp_info_where'] ) ? $display_opp_where = true : $display_opp_where = false;
+      isset( $instance['opp_info_spots'] ) ? $display_opp_spots = true: $display_opp_spots = false;
+
+      // Store number of opps to show in $num_of_opps
+      if ( isset( $instance['number_of_opps_input'] ) ) {
+         $num_of_opps = (int) esc_attr( $instance['number_of_opps_input'] );
+      }
+
+      $list_type = $instance['list_type_radio_btn'];
+
+      // If list type is flexible query for flexible opps else query for one-time opps
+      if( $list_type === 'flexible' ) {
+
+         // Query for flexible volunteer opportunities
+         $posts_args = array(
+            'posts_per_page' => $num_of_opps,
+            'post_type' => 'volunteer_opp',
+            'post_status' => 'publish',
+            'meta_query' => array(
+               array(
+               'key' => '_one_time_opp',
+               'value' => 1,
+               'compare' => '!='
+               )
+            )
+         );
+
+      } else {
+
+         // Query for one-time volunteer opportunities
+         $posts_args = array(
+            'posts_per_page' => $num_of_opps,
+            'post_type' => 'volunteer_opp',
+            'post_status' => 'publish',
+            'meta_key' => '_start_date_time',
+            'orderby' => 'meta_value_num',
+            'order'   => 'ASC',
+            'meta_query' => array(
+               array( //Only if one-time opp is true
+               'key'     => '_one_time_opp',
+               'value'   => 1,
+               'compare' => '==',
+               ),
+               array( //Only if event is in the future
+               'key'     => '_start_date_time',
+               'value'   => current_time( 'timestamp' ),
+               'compare' => '>=',
+               ),
+            'relation' => 'AND'
+            )
+         );
+
+      }
+
+      $opps_query = new WP_Query( $posts_args );
+
+      $template_loader = new WI_Volunteer_Management_Template_Loader(); ?>
+
+      <aside id="opportunities" class="widget widget_links volunteer-opp-info">
```

Comment:
Let's prefix the id with "wivm" so it reads "wivm-opportunities". That way there's very little likelihood someone targets it on accident.